## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#377c4509-3dbe-4cec-b327-5a8c43b902a6) |
| mint-gateway-azure.sh | :x: [more...](#db031064-8d3b-4b24-be65-c95c757d45b1) |
| mint-gateway-nas.sh | :x: [more...](#90483f1a-50d5-4e8b-bc06-3af1232ef485) |
<a name="90483f1a-50d5-4e8b-bc06-3af1232ef485"></a>
### 9442-97710ad/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30296
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6c3b6ed8676c:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 6 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 18 seconds
(8/15) Running minio-dotnet tests ... done in 27 seconds
(9/15) Running minio-go tests ... FAILED in 19 seconds
{}
(9/15) Running minio-java tests ... FAILED in 38 seconds
{
  "name": "minio-java",
  "function": "enableObjectLegalHold(String bucketName, String objectName, String versionId)",
  "duration": 31,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidRequest, message = Bucket is missing ObjectLockConfiguration, bucketName = minio-java-test-38bfica, objectName = minio-java-test-3qf41li, resource = /minio-java-test-38bfica/minio-java-test-3qf41li, requestId = 160A01E1903035B0, hostId = 37cfd697-eb9c-41f6-a1e6-484700908e36)\nrequest={method=PUT, url=http://minio-dev7.minio.io:30296/minio-java-test-38bfica/minio-java-test-3qf41li?legal-hold=, headers=Host: minio-dev7.minio.io:30296\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: erOisiAmIIzsh/Ze2QqGXw==\nx-amz-content-sha256: 96b73c95a8d33e664ab2170e095025b47ebd55978bb71cebd6a51e394bf96722\nx-amz-date: 20200428T142446Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200428/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 384\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 160A01E1903035B0\nX-Xss-Protection: 1; mode=block\nDate: Tue, 28 Apr 2020 14:24:46 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.executePut(MinioClient.java:1424), io.minio.MinioClient.enableObjectLegalHold(MinioClient.java:3568), FunctionalTest.enableObjectLegalHold_test(FunctionalTest.java:3303), FunctionalTest.runTests(FunctionalTest.java:4079), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 35 seconds
(10/15) Running minio-py tests ... done in 59 seconds
(11/15) Running s3cmd tests ... done in 19 seconds
(12/15) Running s3select tests ... done in 1 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="377c4509-3dbe-4cec-b327-5a8c43b902a6"></a>
### 9442-97710ad/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c4.minio.io:31509
ACCESS_KEY:           AKIA3WNQJCXE3JVMGFPN
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5558e80359c7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-n65koxvyjx6xix",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 676,
  "error": "TooManyBuckets: You have attempted to create more buckets than allowed\n\tstatus code: 400, request id: 160A01AFDCAC7CD3, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-n65koxvyjx6xix",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 676,
  "error": "TooManyBuckets: You have attempted to create more buckets than allowed\n\tstatus code: 400, request id: 160A01AFDCAC7CD3, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-n65koxvyjx6xix",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 676,
  "error": "TooManyBuckets: You have attempted to create more buckets than allowed\n\tstatus code: 400, request id: 160A01AFDCAC7CD3, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "makeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-83dd743c21f2"
  },
  "duration": 668.8,
  "error": "You have attempted to create more buckets than allowed",
  "status": "FAIL"
}
(1/15) Running awscli tests ... FAILED in 3 seconds
{
  "name": "awscli",
  "duration": 1762,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (TooManyBuckets) when calling the CreateBucket operation: You have attempted to create more buckets than allowed"
}
(1/15) Running healthcheck tests ... done in 1 seconds
(2/15) Running mc tests ... FAILED in 1 seconds
{
  "name": "mc",
  "duration": "510",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-3083 --quiet --no-color mb myminio/mc-test-bucket-7622  >>> \nmc: <ERROR> Unable to make bucket `myminio/mc-test-bucket-7622`. You have attempted to create more buckets than allowed"
}
(2/15) Running minio-dotnet tests ... done in 25 seconds
(3/15) Running minio-go tests ... FAILED in 1 seconds
{
  "args": {
    "bucketName": "minio-go-test-oi5wkvama5s4eyda",
    "region": "eu-west-1"
  },
  "duration": 421,
  "error": "You have attempted to create more buckets than allowed",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(String bucketName)",
  "duration": 768,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = TooManyBuckets, message = You have attempted to create more buckets than allowed, bucketName = minio-java-test-21mbodh, objectName = null, resource = /minio-java-test-21mbodh, requestId = 160A01B859C81F5A, hostId = d222cf6a-7f3b-4641-aa7d-29d2a853d815)\nrequest={method=PUT, url=http://minio-c4.minio.io:31509/minio-java-test-21mbodh, headers=Host: minio-c4.minio.io:31509\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200428T142149Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200428/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 339\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 160A01B859C81F5A\nX-Xss-Protection: 1; mode=block\nDate: Tue, 28 Apr 2020 14:21:50 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.makeBucket_test1(FunctionalTest.java:247), FunctionalTest.runTests(FunctionalTest.java:4005), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(3/15) Running minio-js tests ... FAILED in 2 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 373,
  "status": "FAIL",
  "error": "S3Error: The specified bucket does not exist at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(3/15) Running minio-py tests ... FAILED in 0 seconds
{
  "name": "minio-py:test_make_bucket_default_region",
  "function": "make_bucket(bucket_name, location)",
  "args": {
    "bucket_name": "minio-py-test-55322816-8d7a-4248-b2bf-0caabac330ee",
    "location": "default value ('us-east-1')"
  },
  "duration": 151,
  "message": "TooManyBuckets: message: You have attempted to create more buckets than allowed.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 188, in test_make_bucket_default_region\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 374, in make_bucket\n    raise ResponseError(response, method, bucket_name).get_exception()\nminio.error.TooManyBuckets: TooManyBuckets: message: You have attempted to create more buckets than allowed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1968, in main\n    test_make_bucket_default_region(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 196, in test_make_bucket_default_region\n    raise Exception(err)\nException: TooManyBuckets: message: You have attempted to create more buckets than allowed.\n",
  "status": "FAIL"
}
(3/15) Running s3cmd tests ... FAILED in 1 seconds
{
  "name": "s3cmd",
  "duration": "665",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "ERROR: S3 error: 400 (TooManyBuckets): You have attempted to create more buckets than allowed"
}
(3/15) Running s3select tests ... FAILED in 2 seconds
{
  "name": "minio-py:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-a412d7ea-cdbe-43d3-95eb-5e651a86d4a3"
  },
  "duration": 1274,
  "message": "NoSuchBucket: message: The specified bucket does not exist.",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 144, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 374, in make_bucket\n    raise ResponseError(response, method, bucket_name).get_exception()\nminio.error.TooManyBuckets: TooManyBuckets: message: You have attempted to create more buckets than allowed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 183, in test_csv_input_custom_quote_char\n    raise Exception(err)\nException: TooManyBuckets: message: You have attempted to create more buckets than allowed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 186, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 446, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1984, in _url_open\n    region = self._get_bucket_region(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1919, in _get_bucket_region\n    region = self._get_bucket_location(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1957, in _get_bucket_location\n    raise ResponseError(response, method, bucket_name).get_exception()\nminio.error.NoSuchBucket: NoSuchBucket: message: The specified bucket does not exist.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 310, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"./tests.py\", line 188, in test_csv_input_custom_quote_char\n    raise Exception(err)\nException: NoSuchBucket: message: The specified bucket does not exist.\n",
  "status": "FAIL"
}
(3/15) Running security tests ... done in 0 seconds

Executed 3 out of 15 tests successfully.
```
<a name="db031064-8d3b-4b24-be65-c95c757d45b1"></a>
### 9442-97710ad/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c4.minio.io:30902
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6dc645d50c7b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 21 seconds
(4/15) Running aws-sdk-ruby tests ... done in 16 seconds
(5/15) Running awscli tests ... FAILED in 47 seconds
{
  "name": "awscli",
  "duration": 3929,
  "function": "aws --endpoint-url http://minio-c4.minio.io:30902 s3api copy-object --bucket awscli-mint-test-bucket-7838 --storage-class REDUCED_REDUNDANCY --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-7838/datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (NotImplemented) when calling the CopyObject operation: A header you provided implies functionality that is not implemented"
}
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... done in 3 minutes and 13 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 23 seconds
(8/15) Running minio-go tests ... FAILED in 3 minutes and 57 seconds
{}
(8/15) Running minio-java tests ... FAILED in 25 minutes and 42 seconds
{
  "name": "minio-java",
  "function": "enableObjectLegalHold(String bucketName, String objectName, String versionId)",
  "duration": 1724,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InvalidRequest, message = Bucket is missing ObjectLockConfiguration, bucketName = minio-java-test-22vus6n, objectName = minio-java-test-29kd335, resource = /minio-java-test-22vus6n/minio-java-test-29kd335, requestId = 160A03C00A1A19B8, hostId = 67d3c52a-3b54-4928-8f82-18d1b3fac0a1)\nrequest={method=PUT, url=http://minio-c4.minio.io:30902/minio-java-test-22vus6n/minio-java-test-29kd335?legal-hold=, headers=Host: minio-c4.minio.io:30902\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: erOisiAmIIzsh/Ze2QqGXw==\nx-amz-content-sha256: 96b73c95a8d33e664ab2170e095025b47ebd55978bb71cebd6a51e394bf96722\nx-amz-date: 20200428T145901Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200428/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=400, headers=Accept-Ranges: bytes\nContent-Length: 384\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 160A03C00A1A19B8\nX-Xss-Protection: 1; mode=block\nDate: Tue, 28 Apr 2020 14:59:02 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.executePut(MinioClient.java:1424), io.minio.MinioClient.enableObjectLegalHold(MinioClient.java:3568), FunctionalTest.enableObjectLegalHold_test(FunctionalTest.java:3303), FunctionalTest.runTests(FunctionalTest.java:4079), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(8/15) Running minio-js tests ... FAILED in 2 minutes and 7 seconds
{
  "name": "minio-js",
  "function": "listObjectsV2WithMetadata(bucketName, prefix, recursive, startAfter)",
  "args": "bucketName:minio-js-test-66679973-dad0-477c-8055-a579c639b846, recursive:true",
  "duration": 209,
  "status": "FAIL",
  "error": "TypeError: Cannot read property '0' of undefined at /mint/run/core/minio-js/node_modules/minio/dist/main/xml-parsers.js:414:42 at Array.forEach (<anonymous>) at parseListObjectsV2WithMetadata (node_modules/minio/dist/main/xml-parsers.js:409:21) at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:88:19) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(8/15) Running minio-py tests ... done in 9 minutes and 37 seconds
(9/15) Running s3cmd tests ... FAILED in 44 seconds
{
  "name": "s3cmd",
  "duration": "1715",
  "function": "test_copy_object",
  "status": "FAIL",
  "error": ""
}
(9/15) Running s3select tests ... done in 22 seconds
(10/15) Running security tests ... done in 0 seconds

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/pkg/bucket/policy/effect.go
+++ b/pkg/bucket/policy/effect.go
@@ -16,10 +16,6 @@
 
 package policy
 
-import (
-	"encoding/json"
-)
-
 // Effect - policy statement effect Allow or Deny.
 type Effect string
 
@@ -48,30 +44,4 @@ func (effect Effect) IsValid() bool {
 	}
 
 	return false
-}
-
-// MarshalJSON - encodes Effect to JSON data.
-func (effect Effect) MarshalJSON() ([]byte, error) {
-	if !effect.IsValid() {
-		return nil, Errorf("invalid effect '%v'", effect)
-	}
-
-	return json.Marshal(string(effect))
-}
-
-// UnmarshalJSON - decodes JSON data to Effect.
-func (effect *Effect) UnmarshalJSON(data []byte) error {
-	var s string
-	if err := json.Unmarshal(data, &s); err != nil {
-		return err
-	}
-
-	e := Effect(s)
-	if !e.IsValid() {
-		return Errorf("invalid effect '%v'", s)
-	}
-
-	*effect = e
-
-	return nil
 }

--- a/pkg/bucket/policy/effect_test.go
+++ b/pkg/bucket/policy/effect_test.go
@@ -17,8 +17,6 @@
 package policy
 
 import (
-	"encoding/json"
-	"reflect"
 	"testing"
 )
 
@@ -60,63 +58,6 @@ func TestEffectIsValid(t *testing.T) {
 
 		if result != testCase.expectedResult {
 			t.Fatalf("case %v: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
-		}
-	}
-}
-
-func TestEffectMarshalJSON(t *testing.T) {
-	testCases := []struct {
-		effect         Effect
-		expectedResult []byte
-		expectErr      bool
-	}{
-		{Allow, []byte(`"Allow"`), false},
-		{Deny, []byte(`"Deny"`), false},
-		{Effect(""), nil, true},
-		{Effect("foo"), nil, true},
-	}
-
-	for i, testCase := range testCases {
-		result, err := json.Marshal(testCase.effect)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v\n", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if !reflect.DeepEqual(result, testCase.expectedResult) {
-				t.Fatalf("case %v: result: expected: %v, got: %v\n", i+1, string(testCase.expectedResult), string(result))
-			}
-		}
-	}
-}
-
-func TestEffectUnmarshalJSON(t *testing.T) {
-	testCases := []struct {
-		data           []byte
-		expectedResult Effect
-		expectErr      bool
-	}{
-		{[]byte(`"Allow"`), Allow, false},
-		{[]byte(`"Deny"`), Deny, false},
-		{[]byte(`""`), Effect(""), true},
-		{[]byte(`"foo"`), Effect(""), true},
-	}
-
-	for i, testCase := range testCases {
-		var result Effect
-		err := json.Unmarshal(testCase.data, &result)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v\n", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if result != testCase.expectedResult {
-				t.Fatalf("case %v: result: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
-			}
 		}
 	}
 }

--- a/pkg/bucket/policy/id.go
+++ b/pkg/bucket/policy/id.go
@@ -17,7 +17,6 @@
 package policy
 
 import (
-	"encoding/json"
 	"unicode/utf8"
 )
 
@@ -27,30 +26,4 @@ type ID string
 // IsValid - checks if ID is valid or not.
 func (id ID) IsValid() bool {
 	return utf8.ValidString(string(id))
-}
-
-// MarshalJSON - encodes ID to JSON data.
-func (id ID) MarshalJSON() ([]byte, error) {
-	if !id.IsValid() {
-		return nil, Errorf("invalid ID %v", id)
-	}
-
-	return json.Marshal(string(id))
-}
-
-// UnmarshalJSON - decodes JSON data to ID.
-func (id *ID) UnmarshalJSON(data []byte) error {
-	var s string
-	if err := json.Unmarshal(data, &s); err != nil {
-		return err
-	}
-
-	i := ID(s)
-	if !i.IsValid() {
-		return Errorf("invalid ID %v", s)
-	}
-
-	*id = i
-
-	return nil
 }

--- a/pkg/bucket/policy/id_test.go
+++ b/pkg/bucket/policy/id_test.go
@@ -17,8 +17,6 @@
 package policy
 
 import (
-	"encoding/json"
-	"reflect"
 	"testing"
 )
 
@@ -37,61 +35,6 @@ func TestIDIsValid(t *testing.T) {
 
 		if result != testCase.expectedResult {
 			t.Errorf("case %v: result: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
-		}
-	}
-}
-
-func TestIDMarshalJSON(t *testing.T) {
-	testCases := []struct {
-		id             ID
-		expectedResult []byte
-		expectErr      bool
-	}{
-		{ID("DenyEncryptionSt1"), []byte(`"DenyEncryptionSt1"`), false},
-		{ID(""), []byte(`""`), false},
-		{ID("aa\xe2"), nil, true}, // invalid utf-8
-	}
-
-	for i, testCase := range testCases {
-		result, err := json.Marshal(testCase.id)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v\n", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if !reflect.DeepEqual(result, testCase.expectedResult) {
-				t.Fatalf("case %v: result: expected: %v, got: %v\n", i+1, string(testCase.expectedResult), string(result))
-			}
-		}
-	}
-}
-
-func TestIDUnmarshalJSON(t *testing.T) {
-	testCases := []struct {
-		data           []byte
-		expectedResult ID
-		expectErr      bool
-	}{
-		{[]byte(`"DenyEncryptionSt1"`), ID("DenyEncryptionSt1"), false},
-		{[]byte(`""`), ID(""), false},
-		{[]byte(`"aa\xe2"`), ID(""), true},
-	}
-
-	for i, testCase := range testCases {
-		var result ID
-		err := json.Unmarshal(testCase.data, &result)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v\n", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if result != testCase.expectedResult {
-				t.Fatalf("case %v: result: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
-			}
 		}
 	}
 }

--- a/pkg/iam/policy/action.go
+++ b/pkg/iam/policy/action.go
@@ -17,8 +17,6 @@
 package iampolicy
 
 import (
-	"encoding/json"
-
 	"github.com/minio/minio/pkg/bucket/policy/condition"
 	"github.com/minio/minio/pkg/wildcard"
 )
@@ -205,47 +203,6 @@ func (action Action) Match(a Action) bool {
 func (action Action) IsValid() bool {
 	_, ok := supportedActions[action]
 	return ok
-}
-
-// MarshalJSON - encodes Action to JSON data.
-func (action Action) MarshalJSON() ([]byte, error) {
-	if action.IsValid() || AdminAction(action).IsValid() {
-		return json.Marshal(string(action))
-	}
-
-	return nil, Errorf("invalid action '%v'", action)
-}
-
-// UnmarshalJSON - decodes JSON data to Action.
-func (action *Action) UnmarshalJSON(data []byte) error {
-	var s string
-
-	if err := json.Unmarshal(data, &s); err != nil {
-		return err
-	}
-
-	a := Action(s)
-	if !a.IsValid() {
-		return Errorf("invalid action '%v'", s)
-	}
-
-	*action = a
-
-	return nil
-}
-
-func parseAction(s string) (Action, error) {
-	adminAction, err := parseAdminAction(s)
-	if err == nil {
-		return Action(adminAction), nil
-	}
-	action := Action(s)
-
-	if action.IsValid() {
-		return action, nil
-	}
-
-	return action, Errorf("unsupported action '%v'", s)
 }
 
 // actionConditionKeyMap - holds mapping of supported condition key for an action.

--- a/pkg/iam/policy/action_test.go
+++ b/pkg/iam/policy/action_test.go
@@ -17,8 +17,6 @@
 package iampolicy
 
 import (
-	"encoding/json"
-	"reflect"
 	"testing"
 )
 
@@ -49,6 +47,7 @@ func TestActionIsValid(t *testing.T) {
 		action         Action
 		expectedResult bool
 	}{
+		{PutObjectAction, true},
 		{AbortMultipartUploadAction, true},
 		{Action("foo"), false},
 	}
@@ -58,59 +57,6 @@ func TestActionIsValid(t *testing.T) {
 
 		if testCase.expectedResult != result {
 			t.Fatalf("case %v: expected: %v, got: %v", i+1, testCase.expectedResult, result)
-		}
-	}
-}
-
-func TestActionMarshalJSON(t *testing.T) {
-	testCases := []struct {
-		action         Action
-		expectedResult []byte
-		expectErr      bool
-	}{
-		{PutObjectAction, []byte(`"s3:PutObject"`), false},
-		{Action("foo"), nil, true},
-	}
-
-	for i, testCase := range testCases {
-		result, err := json.Marshal(testCase.action)
-		expectErr := (err != nil)
-
-		if testCase.expectErr != expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if !reflect.DeepEqual(result, testCase.expectedResult) {
-				t.Fatalf("case %v: result: expected: %v, got: %v", i+1, testCase.expectedResult, result)
-			}
-		}
-	}
-}
-
-func TestActionUnmarshalJSON(t *testing.T) {
-	testCases := []struct {
-		data           []byte
-		expectedResult Action
-		expectErr      bool
-	}{
-		{[]byte(`"s3:PutObject"`), PutObjectAction, false},
-		{[]byte(`"foo"`), Action(""), true},
-	}
-
-	for i, testCase := range testCases {
-		var result Action
-		err := json.Unmarshal(testCase.data, &result)
-		expectErr := (err != nil)
-
-		if testCase.expectErr != expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if testCase.expectedResult != result {
-				t.Fatalf("case %v: result: expected: %v, got: %v", i+1, testCase.expectedResult, result)
-			}
 		}
 	}
 }

--- a/pkg/iam/policy/actionset.go
+++ b/pkg/iam/policy/actionset.go
@@ -107,6 +107,16 @@ func (actionSet ActionSet) ToSlice() []Action {
 	return actions
 }
 
+// ToAdminSlice - returns slice of admin actions from the action set.
+func (actionSet ActionSet) ToAdminSlice() []AdminAction {
+	actions := []AdminAction{}
+	for action := range actionSet {
+		actions = append(actions, AdminAction(action))
+	}
+
+	return actions
+}
+
 // UnmarshalJSON - decodes JSON data to ActionSet.
 func (actionSet *ActionSet) UnmarshalJSON(data []byte) error {
 	var sset set.StringSet
@@ -120,14 +130,29 @@ func (actionSet *ActionSet) UnmarshalJSON(data []byte) error {
 
 	*actionSet = make(ActionSet)
 	for _, s := range sset.ToSlice() {
-		action, err := parseAction(s)
-		if err != nil {
-			return err
-		}
-
-		actionSet.Add(action)
+		actionSet.Add(Action(s))
 	}
 
+	return nil
+}
+
+// ValidateAdmin checks if all actions are valid Admin actions
+func (actionSet ActionSet) ValidateAdmin() error {
+	for _, action := range actionSet.ToAdminSlice() {
+		if !action.IsValid() {
+			return Errorf("unsupported admin action '%v'", action)
+		}
+	}
+	return nil
+}
+
+// Validate checks if all actions are valid
+func (actionSet ActionSet) Validate() error {
+	for _, action := range actionSet.ToSlice() {
+		if !action.IsValid() {
+			return Errorf("unsupported action '%v'", action)
+		}
+	}
 	return nil
 }
 

--- a/pkg/iam/policy/actionset_test.go
+++ b/pkg/iam/policy/actionset_test.go
@@ -128,17 +128,18 @@ func TestActionSetToSlice(t *testing.T) {
 
 func TestActionSetUnmarshalJSON(t *testing.T) {
 	testCases := []struct {
-		data           []byte
-		expectedResult ActionSet
-		expectErr      bool
+		data               []byte
+		expectedResult     ActionSet
+		expectUnmarshalErr bool
+		expectValidateErr  bool
 	}{
-		{[]byte(`"s3:PutObject"`), NewActionSet(PutObjectAction), false},
-		{[]byte(`["s3:PutObject"]`), NewActionSet(PutObjectAction), false},
-		{[]byte(`["s3:PutObject", "s3:GetObject"]`), NewActionSet(PutObjectAction, GetObjectAction), false},
-		{[]byte(`["s3:PutObject", "s3:GetObject", "s3:PutObject"]`), NewActionSet(PutObjectAction, GetObjectAction), false},
-		{[]byte(`[]`), NewActionSet(), true},           // Empty array.
-		{[]byte(`"foo"`), nil, true},                   // Invalid action.
-		{[]byte(`["s3:PutObject", "foo"]`), nil, true}, // Invalid action.
+		{[]byte(`"s3:PutObject"`), NewActionSet(PutObjectAction), false, false},
+		{[]byte(`["s3:PutObject"]`), NewActionSet(PutObjectAction), false, false},
+		{[]byte(`["s3:PutObject", "s3:GetObject"]`), NewActionSet(PutObjectAction, GetObjectAction), false, false},
+		{[]byte(`["s3:PutObject", "s3:GetObject", "s3:PutObject"]`), NewActionSet(PutObjectAction, GetObjectAction), false, false},
+		{[]byte(`[]`), NewActionSet(), true, false},           // Empty array.
+		{[]byte(`"foo"`), nil, false, true},                   // Invalid action.
+		{[]byte(`["s3:PutObject", "foo"]`), nil, false, true}, // Invalid action.
 	}
 
 	for i, testCase := range testCases {
@@ -146,11 +147,17 @@ func TestActionSetUnmarshalJSON(t *testing.T) {
 		err := json.Unmarshal(testCase.data, &result)
 		expectErr := (err != nil)
 
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v\n", i+1, testCase.expectErr, expectErr)
+		if expectErr != testCase.expectUnmarshalErr {
+			t.Fatalf("case %v: error during unmarshal: expected: %v, got: %v\n", i+1, testCase.expectUnmarshalErr, expectErr)
 		}
 
-		if !testCase.expectErr {
+		err = result.Validate()
+		expectErr = (err != nil)
+		if expectErr != testCase.expectValidateErr {
+			t.Fatalf("case %v: error during validation: expected: %v, got: %v\n", i+1, testCase.expectValidateErr, expectErr)
+		}
+
+		if !testCase.expectUnmarshalErr && !testCase.expectValidateErr {
 			if !reflect.DeepEqual(result, testCase.expectedResult) {
 				t.Fatalf("case %v: result: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
 			}

--- a/pkg/iam/policy/policy.go
+++ b/pkg/iam/policy/policy.go
@@ -103,19 +103,7 @@ func (iamp Policy) isValid() error {
 			return err
 		}
 	}
-
 	return nil
-}
-
-// MarshalJSON - encodes Policy to JSON data.
-func (iamp Policy) MarshalJSON() ([]byte, error) {
-	if err := iamp.isValid(); err != nil {
-		return nil, err
-	}
-
-	// subtype to avoid recursive call to MarshalJSON()
-	type subPolicy Policy
-	return json.Marshal(subPolicy(iamp))
 }
 
 func (iamp *Policy) dropDuplicateStatements() {
@@ -153,30 +141,14 @@ func (iamp *Policy) UnmarshalJSON(data []byte) error {
 	}
 
 	p := Policy(sp)
-	if err := p.isValid(); err != nil {
-		return err
-	}
-
 	p.dropDuplicateStatements()
-
 	*iamp = p
-
 	return nil
 }
 
 // Validate - validates all statements are for given bucket or not.
 func (iamp Policy) Validate() error {
-	if err := iamp.isValid(); err != nil {
-		return err
-	}
-
-	for _, statement := range iamp.Statements {
-		if err := statement.Validate(); err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return iamp.isValid()
 }
 
 // ParseConfig - parses data in given reader to Iamp.

--- a/pkg/iam/policy/policy_test.go
+++ b/pkg/iam/policy/policy_test.go
@@ -387,226 +387,7 @@ func TestPolicyIsValid(t *testing.T) {
 	}
 }
 
-func TestPolicyMarshalJSON(t *testing.T) {
-	case1Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case1Policy.Statements[0].SID = "SomeId1"
-	case1Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Sid":"SomeId1","Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]}]}`)
-
-	_, IPNet1, err := net.ParseCIDR("192.168.1.0/24")
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-	func1, err := condition.NewIPAddressFunc(
-		condition.AWSSourceIP,
-		IPNet1,
-	)
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-
-	case2Policy := Policy{
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-			NewStatement(
-				policy.Deny,
-				NewActionSet(GetObjectAction),
-				NewResourceSet(NewResource("mybucket", "/yourobject*")),
-				condition.NewFunctions(func1),
-			),
-		},
-	}
-	case2Data := []byte(`{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]},{"Effect":"Deny","Action":["s3:GetObject"],"Resource":["arn:aws:s3:::mybucket/yourobject*"],"Condition":{"IpAddress":{"aws:SourceIp":["192.168.1.0/24"]}}}]}`)
-
-	case3Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(GetObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case3Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:GetObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]},{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]}]}`)
-
-	case4Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-			NewStatement(
-				policy.Allow,
-				NewActionSet(GetObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case4Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]},{"Effect":"Allow","Action":["s3:GetObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]}]}`)
-
-	case5Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(),
-			),
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/yourobject*")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case5Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]},{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/yourobject*"]}]}`)
-
-	_, IPNet2, err := net.ParseCIDR("192.168.2.0/24")
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-	func2, err := condition.NewIPAddressFunc(
-		condition.AWSSourceIP,
-		IPNet2,
-	)
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-
-	case6Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(func1),
-			),
-			NewStatement(
-				policy.Allow,
-				NewActionSet(PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "/myobject*")),
-				condition.NewFunctions(func2),
-			),
-		},
-	}
-	case6Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"],"Condition":{"IpAddress":{"aws:SourceIp":["192.168.1.0/24"]}}},{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"],"Condition":{"IpAddress":{"aws:SourceIp":["192.168.2.0/24"]}}}]}`)
-
-	case7Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(GetBucketLocationAction),
-				NewResourceSet(NewResource("mybucket", "")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case7Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:GetBucketLocation"],"Resource":["arn:aws:s3:::mybucket"]}]}`)
-
-	case8Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(GetBucketLocationAction),
-				NewResourceSet(NewResource("*", "")),
-				condition.NewFunctions(),
-			),
-		},
-	}
-	case8Data := []byte(`{"ID":"MyPolicyForMyBucket1","Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:GetBucketLocation"],"Resource":["arn:aws:s3:::*"]}]}`)
-
-	func3, err := condition.NewNullFunc(
-		condition.S3XAmzCopySource,
-		true,
-	)
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-	case9Policy := Policy{
-		ID:      "MyPolicyForMyBucket1",
-		Version: DefaultVersion,
-		Statements: []Statement{
-			NewStatement(
-				policy.Allow,
-				NewActionSet(GetObjectAction, PutObjectAction),
-				NewResourceSet(NewResource("mybucket", "myobject*")),
-				condition.NewFunctions(func1, func2, func3),
-			),
-		},
-	}
-
-	testCases := []struct {
-		policy         Policy
-		expectedResult []byte
-		expectErr      bool
-	}{
-		{case1Policy, case1Data, false},
-		{case2Policy, case2Data, false},
-		{case3Policy, case3Data, false},
-		{case4Policy, case4Data, false},
-		{case5Policy, case5Data, false},
-		{case6Policy, case6Data, false},
-		{case7Policy, case7Data, false},
-		{case8Policy, case8Data, false},
-		{case9Policy, nil, true},
-	}
-
-	for i, testCase := range testCases {
-		result, err := json.Marshal(testCase.policy)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if !reflect.DeepEqual(result, testCase.expectedResult) {
-				t.Fatalf("case %v: result: expected: %v, got: %v", i+1, string(testCase.expectedResult), string(result))
-			}
-		}
-	}
-}
-
-func TestPolicyUnmarshalJSON(t *testing.T) {
+func TestPolicyUnmarshalJSONAndValidate(t *testing.T) {
 	case1Data := []byte(`{
     "ID": "MyPolicyForMyBucket1",
     "Version": "2012-10-17",
@@ -973,24 +754,25 @@ func TestPolicyUnmarshalJSON(t *testing.T) {
 	}
 
 	testCases := []struct {
-		data           []byte
-		expectedResult Policy
-		expectErr      bool
+		data                []byte
+		expectedResult      Policy
+		expectUnmarshalErr  bool
+		expectValidationErr bool
 	}{
-		{case1Data, case1Policy, false},
-		{case2Data, case2Policy, false},
-		{case3Data, case3Policy, false},
-		{case4Data, case4Policy, false},
-		{case5Data, case5Policy, false},
-		{case6Data, case6Policy, false},
-		{case7Data, case7Policy, false},
-		{case8Data, case8Policy, false},
+		{case1Data, case1Policy, false, false},
+		{case2Data, case2Policy, false, false},
+		{case3Data, case3Policy, false, false},
+		{case4Data, case4Policy, false, false},
+		{case5Data, case5Policy, false, false},
+		{case6Data, case6Policy, false, false},
+		{case7Data, case7Policy, false, false},
+		{case8Data, case8Policy, false, false},
 		// Invalid version error.
-		{case9Data, Policy{}, true},
+		{case9Data, Policy{}, false, true},
 		// Duplicate statement success, duplicate statement is removed.
-		{case10Data, case10Policy, false},
+		{case10Data, case10Policy, false, false},
 		// Duplicate statement success (Effect differs).
-		{case11Data, case11Policy, false},
+		{case11Data, case11Policy, false, false},
 	}
 
 	for i, testCase := range testCases {
@@ -998,11 +780,18 @@ func TestPolicyUnmarshalJSON(t *testing.T) {
 		err := json.Unmarshal(testCase.data, &result)
 		expectErr := (err != nil)
 
-		if expectErr != testCase.expectErr {
-			t.Errorf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
+		if expectErr != testCase.expectUnmarshalErr {
+			t.Errorf("case %v: error during unmarshal: expected: %v, got: %v", i+1, testCase.expectUnmarshalErr, expectErr)
 		}
 
-		if !testCase.expectErr {
+		err = result.Validate()
+		expectErr = (err != nil)
+
+		if expectErr != testCase.expectValidationErr {
+			t.Errorf("case %v: error during validation: expected: %v, got: %v", i+1, testCase.expectValidationErr, expectErr)
+		}
+
+		if !testCase.expectUnmarshalErr && !testCase.expectValidationErr {
 			if !reflect.DeepEqual(result, testCase.expectedResult) {
 				t.Errorf("case %v: result: expected: %v, got: %v", i+1, testCase.expectedResult, result)
 			}

--- a/pkg/iam/policy/statement.go
+++ b/pkg/iam/policy/statement.go
@@ -17,7 +17,6 @@
 package iampolicy
 
 import (
-	"encoding/json"
 	"strings"
 
 	"github.com/minio/minio/pkg/bucket/policy"
@@ -63,11 +62,11 @@ func (statement Statement) IsAllowed(args Args) bool {
 }
 func (statement Statement) isAdmin() bool {
 	for action := range statement.Actions {
-		if !AdminAction(action).IsValid() {
-			return false
+		if AdminAction(action).IsValid() {
+			return true
 		}
 	}
-	return true
+	return false
 }
 
 // isValid - checks whether statement is valid or not.
@@ -81,6 +80,9 @@ func (statement Statement) isValid() error {
 	}
 
 	if statement.isAdmin() {
+		if err := statement.Actions.ValidateAdmin(); err != nil {
+			return err
+		}
 		for action := range statement.Actions {
 			keys := statement.Conditions.Keys()
 			keyDiff := keys.Difference(adminActionConditionKeyMap[action])
@@ -91,11 +93,19 @@ func (statement Statement) isValid() error {
 		return nil
 	}
 
+	if !statement.SID.IsValid() {
+		return Errorf("invalid SID %v", statement.SID)
+	}
+
 	if len(statement.Resources) == 0 {
 		return Errorf("Resource must not be empty")
 	}
 
 	if err := statement.Resources.Validate(); err != nil {
+		return err
+	}
+
+	if err := statement.Actions.Validate(); err != nil {
 		return err
 	}
 
@@ -110,38 +120,6 @@ func (statement Statement) isValid() error {
 			return Errorf("unsupported condition keys '%v' used for action '%v'", keyDiff, action)
 		}
 	}
-
-	return nil
-}
-
-// MarshalJSON - encodes JSON data to Statement.
-func (statement Statement) MarshalJSON() ([]byte, error) {
-	if err := statement.isValid(); err != nil {
-		return nil, err
-	}
-
-	// subtype to avoid recursive call to MarshalJSON()
-	type subStatement Statement
-	ss := subStatement(statement)
-	return json.Marshal(ss)
-}
-
-// UnmarshalJSON - decodes JSON data to Statement.
-func (statement *Statement) UnmarshalJSON(data []byte) error {
-	// subtype to avoid recursive call to UnmarshalJSON()
-	type subStatement Statement
-	var ss subStatement
-
-	if err := json.Unmarshal(data, &ss); err != nil {
-		return err
-	}
-
-	s := Statement(ss)
-	if err := s.isValid(); err != nil {
-		return err
-	}
-
-	*statement = s
 
 	return nil
 }

--- a/pkg/iam/policy/statement_test.go
+++ b/pkg/iam/policy/statement_test.go
@@ -265,82 +265,7 @@ func TestStatementIsValid(t *testing.T) {
 	}
 }
 
-func TestStatementMarshalJSON(t *testing.T) {
-	case1Statement := NewStatement(
-		policy.Allow,
-		NewActionSet(PutObjectAction),
-		NewResourceSet(NewResource("mybucket", "/myobject*")),
-		condition.NewFunctions(),
-	)
-	case1Statement.SID = "SomeId1"
-	case1Data := []byte(`{"Sid":"SomeId1","Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"]}`)
-
-	func1, err := condition.NewNullFunc(
-		condition.S3XAmzCopySource,
-		true,
-	)
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-	case2Statement := NewStatement(
-		policy.Allow,
-		NewActionSet(PutObjectAction),
-		NewResourceSet(NewResource("mybucket", "/myobject*")),
-		condition.NewFunctions(func1),
-	)
-	case2Data := []byte(`{"Effect":"Allow","Action":["s3:PutObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"],"Condition":{"Null":{"s3:x-amz-copy-source":[true]}}}`)
-
-	func2, err := condition.NewNullFunc(
-		condition.S3XAmzServerSideEncryption,
-		false,
-	)
-	if err != nil {
-		t.Fatalf("unexpected error. %v\n", err)
-	}
-	case3Statement := NewStatement(
-		policy.Deny,
-		NewActionSet(GetObjectAction),
-		NewResourceSet(NewResource("mybucket", "/myobject*")),
-		condition.NewFunctions(func2),
-	)
-	case3Data := []byte(`{"Effect":"Deny","Action":["s3:GetObject"],"Resource":["arn:aws:s3:::mybucket/myobject*"],"Condition":{"Null":{"s3:x-amz-server-side-encryption":[false]}}}`)
-
-	case4Statement := NewStatement(
-		policy.Allow,
-		NewActionSet(GetObjectAction, PutObjectAction),
-		NewResourceSet(NewResource("mybucket", "myobject*")),
-		condition.NewFunctions(func1, func2),
-	)
-
-	testCases := []struct {
-		statement      Statement
-		expectedResult []byte
-		expectErr      bool
-	}{
-		{case1Statement, case1Data, false},
-		{case2Statement, case2Data, false},
-		{case3Statement, case3Data, false},
-		// Invalid statement error.
-		{case4Statement, nil, true},
-	}
-
-	for i, testCase := range testCases {
-		result, err := json.Marshal(testCase.statement)
-		expectErr := (err != nil)
-
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
-		}
-
-		if !testCase.expectErr {
-			if !reflect.DeepEqual(result, testCase.expectedResult) {
-				t.Fatalf("case %v: result: expected: %v, got: %v", i+1, string(testCase.expectedResult), string(result))
-			}
-		}
-	}
-}
-
-func TestStatementUnmarshalJSON(t *testing.T) {
+func TestStatementUnmarshalJSONAndValidate(t *testing.T) {
 	case1Data := []byte(`{
     "Sid": "SomeId1",
     "Effect": "Allow",
@@ -408,7 +333,7 @@ func TestStatementUnmarshalJSON(t *testing.T) {
 
 	case4Data := []byte(`{
     "Effect": "Allow",
-    "Action": "s3:PutObjec",
+    "Action": "s3:PutObjec,
     "Resource": "arn:aws:s3:::mybucket/myobject*"
 }`)
 
@@ -450,36 +375,42 @@ func TestStatementUnmarshalJSON(t *testing.T) {
 }`)
 
 	testCases := []struct {
-		data           []byte
-		expectedResult Statement
-		expectErr      bool
+		data                []byte
+		expectedResult      Statement
+		expectUnmarshalErr  bool
+		expectValidationErr bool
 	}{
-		{case1Data, case1Statement, false},
-		{case2Data, case2Statement, false},
-		{case3Data, case3Statement, false},
+		{case1Data, case1Statement, false, false},
+		{case2Data, case2Statement, false, false},
+		{case3Data, case3Statement, false, false},
 		// JSON unmarshaling error.
-		{case4Data, Statement{}, true},
+		{case4Data, Statement{}, true, true},
 		// Invalid effect error.
-		{case5Data, Statement{}, true},
+		{case5Data, Statement{}, false, true},
 		// Empty action error.
-		{case7Data, Statement{}, true},
+		{case7Data, Statement{}, false, true},
 		// Empty resource error.
-		{case8Data, Statement{}, true},
+		{case8Data, Statement{}, false, true},
 		// Empty condition error.
-		{case9Data, Statement{}, true},
+		{case9Data, Statement{}, true, false},
 		// Unsupported condition key error.
-		{case10Data, Statement{}, true},
+		{case10Data, Statement{}, false, true},
 	}
 
 	for i, testCase := range testCases {
 		var result Statement
 		expectErr := (json.Unmarshal(testCase.data, &result) != nil)
 
-		if expectErr != testCase.expectErr {
-			t.Fatalf("case %v: error: expected: %v, got: %v", i+1, testCase.expectErr, expectErr)
+		if expectErr != testCase.expectUnmarshalErr {
+			t.Fatalf("case %v: error during unmarshal: expected: %v, got: %v", i+1, testCase.expectUnmarshalErr, expectErr)
 		}
 
-		if !testCase.expectErr {
+		expectErr = (result.Validate() != nil)
+		if expectErr != testCase.expectValidationErr {
+			t.Fatalf("case %v: error during validation: expected: %v, got: %v", i+1, testCase.expectValidationErr, expectErr)
+		}
+
+		if !testCase.expectUnmarshalErr && !testCase.expectValidationErr {
 			if !reflect.DeepEqual(result, testCase.expectedResult) {
 				t.Fatalf("case %v: result: expected: %v, got: %v", i+1, testCase.expectedResult, result)
 			}

